File: sdkwork-models-typescript/src/validation.ts
import type { ModelCatalog } from "./types.js";

const DECIMAL_PATTERN = /^(0|[1-9][0-9]*)(\.[0-9]+)?$/;

export interface CatalogValidationIssue {
  code: string;
  path: string;
  message: string;
  severity: "error" | "warning";
}

export function validateCatalog(catalog: ModelCatalog): CatalogValidationIssue[] {
  const issues: CatalogValidationIssue[] = [];
  const meters = new Set(catalog.meters.map((meter) => meter.meterCode));
  const models = new Set(catalog.vendors.flatMap((vendor) => vendor.models.map((model) => model.catalogKey)));
  for (const vendor of catalog.vendors) {
    for (const pricing of vendor.pricing) {
      if (!models.has(pricing.catalogKey)) {
        issues.push({
          code: "pricing.model.missing",
          path: pricing.catalogKey,
          message: `${pricing.modelId} is not defined`,
          severity: "error",
        });
      }
      for (const price of pricing.prices) {
        if (!meters.has(price.meterCode)) {
          issues.push({
            code: "pricing.meter.missing",
            path: `${pricing.catalogKey}/${price.meterCode}`,
            message: `${price.meterCode} is not defined`,
            severity: "error",
          });
        }
        for (const field of ["unitSize", "unitPrice", "minimumQuantity"] as const) {
          if (!DECIMAL_PATTERN.test(price[field])) {
            issues.push({
              code: "pricing.decimal.invalid",
              path: `${pricing.catalogKey}/${field}`,
              message: `${field} must be a decimal string`,
              severity: "error",
            });
          }
        }
      }
    }
  }
  return issues;
}

File: sdks/sdkwork-models-sdk/sdkwork-models-sdk-python/sdkwork_models/validation.py
from __future__ import annotations

import re

from .types import ModelCatalog


DECIMAL_PATTERN = re.compile(r"^(0|[1-9][0-9]*)(\.[0-9]+)?$")


def validate_catalog(catalog: ModelCatalog) -> list[dict]:
    meters = {meter["meterCode"] for meter in catalog.meters}
    catalog_keys = {model["catalogKey"] for model in catalog.models}
    issues: list[dict] = []
    for pricing in catalog.pricing:
        if pricing.get("catalogKey") not in catalog_keys:
            issues.append({"code": "pricing.model.missing", "severity": "error", "message": pricing.get("catalogKey")})
        for price in pricing.get("prices", []):
            if price.get("meterCode") not in meters:
                issues.append({"code": "pricing.meter.missing", "severity": "error", "message": price.get("meterCode")})
            for field in ("unitSize", "unitPrice", "minimumQuantity"):
                if not isinstance(price.get(field), str) or not DECIMAL_PATTERN.match(price[field]):
                    issues.append({"code": "pricing.decimal.invalid", "severity": "error", "message": field})
    return issues
